org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/DocumentLinkDetector.java

```diff
@@ -80,35 +80,18 @@ public IHyperlink[] detectHyperlinks(ITextViewer textViewer, IRegion region, boo
 		}
 		final var params = new DocumentLinkParams(LSPEclipseUtils.toTextDocumentIdentifier(uri));
 		try {
-			return LanguageServiceAccessor
-					.getLanguageServers(document,
-							capabilities -> capabilities.getDocumentLinkProvider() != null)
-					.thenApplyAsync(languageServers -> {
-						IHyperlink[] res = languageServers.stream()
-								.map(languageServer -> languageServer.getTextDocumentService().documentLink(params))
-								.map(future -> {
-									try {
-										return future.get(2, TimeUnit.SECONDS);
-									} catch (ExecutionException e) {
-										LanguageServerPlugin.logError(e);
-										return null;
-									} catch (InterruptedException e) {
-										LanguageServerPlugin.logError(e);
-										Thread.currentThread().interrupt();
-										return null;
-									} catch (TimeoutException e) {
-										LanguageServerPlugin.logWarning("Could not detect hyperlinks due to timeout after 2 seconds in `document/Link`", e); //$NON-NLS-1$
-										return null;
-									}
-								}).filter(Objects::nonNull).flatMap(List<DocumentLink>::stream).map(link -> {
+			return LanguageServers.forDocument(document)
+					.withFilter(capabilities -> capabilities.getDocumentLinkProvider() != null)
+					.collectAll(languageServer -> languageServer.getTextDocumentService().documentLink(params))
+					.thenApplyAsync(links -> {
+						IHyperlink[] res = links.stream().flatMap(List<DocumentLink>::stream).filter(Objects::nonNull)
+								.filter(link -> link.getTarget() != null).map(link -> {
```

Comment:
Not your doing, but this is a bit of a monster lambda expression.

Could you pull some of it out into a method while you're in the code? Maybe a (lsp4j) DocumentLink -> (eclipse) DocumentHyperLink converter?